src/__tests__/integration/basic/constructor.test.ts

```diff
@@ -0,0 +1,267 @@
+      await contract.writeInteraction({ function: 'nop' });
+
+      const {
+        cachedValue: { state: state2 }
+      } = await contract.readState();
+      expect(state2.counter).toStrictEqual(2);
+    });
+
+    describe('Internal writes', () => {
+      it('should throw when using internal writes in contract in __init', async () => {
+        const writesInConstructorContract = await deployContract({
+          src: helperContractSrc,
+          withKv: false
+        });
+
+        await expect(writesInConstructorContract.readState()).rejects.toThrowError();
+      });
+
+      it('should read properly from external contract which uses constructor', async () => {
+        const withConstructorContract = await deployContract({
+          src: dummyContractSrc,
+          withKv: false
+        });
+        const readExternalContract = await deployContract({
+          src: contractIrSrc,
+          withKv: false,
+          addToState: { foreignContract: withConstructorContract.txId() }
+        });
+
+        expect((await readExternalContract.viewState({ function: 'read' })).result).toEqual({
+          originalErrorMessages: {},
+          originalValidity: {},
+          result: 100,
+          state: {
+            counter: 100
+          },
+          type: 'ok'
+        });
+      });
+    });
+  });
+
+  describe('with useConstructor = false', () => {
+    describe('0 interactions', () => {
+      it('should not call constructor on first read state', async () => {
+        const contract = await deployContract({ withConstructor: false });
+
+        const {
+          cachedValue: { state }
+        } = await contract.readState();
+
+        expect(state.calls).toBeUndefined();
+
+        const {
+          cachedValue: { state: state2 }
+        } = await contract.readState();
+
+        expect(state2.calls).toBeUndefined();
+      });
+    });
+
+    describe('with missing interactions', () => {
+      it('should not call constructor on first read state and works with next readStates', async () => {
+        const contract = await deployContract({ withConstructor: false });
+
+        await contract.writeInteraction({ function: 'nop' });
+
+        const {
+          cachedValue: { state }
+        } = await contract.readState();
+
+        expect(state.calls).toBeDefined();
+        expect(state.calls).toEqual(['nop']);
+        await contract.writeInteraction({ function: 'nop' });
+        await contract.writeInteraction({ function: 'nop' });
+
+        const {
+          cachedValue: { state: state2 }
+        } = await contract.readState();
+
+        expect(state2.calls).toBeDefined();
+        expect(state2.calls).toEqual(['nop', 'nop', 'nop']);
+      });
+    });
+
+    it('should NOT fail to call __init function', async () => {
+      const contract = await deployContract({ withConstructor: false });
+
+      await expect(contract.writeInteraction({ function: '__init' })).resolves.toBeDefined();
```

Comment:
Here was missing `toBeDefined()` and that is way tests were failing, something specific to jest matchers :)